src/socket/udp_socket.cr

```diff
@@ -88,4 +88,117 @@ class UDPSocket < IPSocket
     bytes_read, sockaddr, addrlen = recvfrom(message)
     {bytes_read, IPAddress.from(sockaddr, addrlen)}
   end
+
+  # Multicast loopback reports whether transmitted multicast packets
+  # should be copied and sent back to the originator.
+  def multicast_loopback?
+    if @family == Family::INET
+      optname = LibC::IP_MULTICAST_LOOP
+      level = LibC::IPPROTO_IP
+    else
+      optname = LibC::IPV6_MULTICAST_LOOP
+      level = LibC::IPPROTO_IPV6
+    end
+    getsockopt_bool optname, level
+  end
+
+  # Multicast loopback sets whether transmitted multicast packets should be
+  # copied and sent back to the originator, if the host has joined the
+  # multicast group.
+  def multicast_loopback=(val : Bool)
+    if @family == Family::INET
+      optname = LibC::IP_MULTICAST_LOOP
+      level = LibC::IPPROTO_IP
+    else
+      optname = LibC::IPV6_MULTICAST_LOOP
+      level = LibC::IPPROTO_IPV6
+    end
+    setsockopt_bool optname, val, level
+  end
+
+  def multicast_hops
+    if @family == Family::INET
+      optname = LibC::IP_MULTICAST_TTL
+      level = LibC::IPPROTO_IP
+    else
+      optname = LibC::IPV6_MULTICAST_HOPS
+      level = LibC::IPPROTO_IPV6
+    end
+    getsockopt optname, 0, level
+  end
+
+  # The multicast hops option controls the `hoplimit` field on uni-cast packets.
+  # If `-1` is specified, the kernel will use a default value.
+  # If a value of `0` to `255` is specified, the packet will have the specified
+  # value as `hoplimit`. Other values are considered invalid.
+  # Datagrams with a `hoplimit` of `1` are not forwarded beyond the local network.
+  # Multicast datagrams with a `hoplimit` of `0` will not be transmitted on any
+  # network, but may be delivered locally if the sending host belongs to the
+  # destination group and multicast loopback is enabled.
+  def multicast_hops=(val : Int)
+    if @family == Family::INET
+      optname = LibC::IP_MULTICAST_TTL
+      level = LibC::IPPROTO_IP
+    else
+      optname = LibC::IPV6_MULTICAST_HOPS
+      level = LibC::IPPROTO_IPV6
+    end
+    setsockopt optname, val, level
+    val
+  end
+
+  # For hosts with multiple interfaces, each multicast transmission is sent
+  # from the primary network interface. This function overrides the default
+  # for subsequent transmissions. Setting the interface to `0.0.0.0` for IPv4
+  # or index `0` for IPv6 will select the default interface.
+  def multicast_interface(ifout : IPAddress | UInt32)
```

Comment:
I disagree with this to some degree. Firstly almost no one is going to use this function, however for those who need it, it should work for both IPv4 and v6.

As ethernet ports would be assigned to VLans which might run DHCP, the IPv6 index method to me makes the most sense. If you were to use this function, you'd be enumerating the network cards on your machine and then setting an ENV var, or equivalent configuration, that would be picked up by the application (it's unlikely to need the application to auto detect anything)

What might be a cleaner solution is to split it up into two functions - as it's the only method who's interfaces are not equivalent between IPv4 and 6

i.e. `ipv4_multicast_interface(ifout : IPAddress)` and `ipv6_multicast_interface(ifout : UInt32)`
But I think this is actually more ugly and anyone using this method should already be aware of the differences between IP implementations

Reply:
I did update this function to accept IPv6 ANY as a valid param for IPv6 though.

Reply:
thoughts @straight-shoota ?

Reply:
Sry, I don't have any decisive opinion on this.
I guess we can keep the `UInt32` argument, but I think it should better be two separate methods, because they're working differently. This would also make it easier to document because the different behaviour is clearly separated.

Reply:
OK. I'll separate them :)